examples/mlflow_deployment/pipeline.py

```diff
@@ -0,0 +1,217 @@
+) -> Output(x_train_normed=np.ndarray, x_test_normed=np.ndarray):
+    """Normalize the values for all the images so they are between 0 and 1"""
+    x_train_normed = x_train / 255.0
+    x_test_normed = x_test / 255.0
+    return x_train_normed, x_test_normed
+
+
+# Define the step and enable mlflow - order of decorators is important here
+@enable_mlflow
+@step
+def tf_trainer(
+    config: TrainerConfig,
+    x_train: np.ndarray,
+    y_train: np.ndarray,
+) -> tf.keras.Model:
+    """Train a neural net from scratch to recognize MNIST digits return our
+    model or the learner"""
+    model = tf.keras.Sequential(
+        [
+            tf.keras.layers.Flatten(input_shape=(28, 28)),
+            tf.keras.layers.Dense(10),
+        ]
+    )
+
+    model.compile(
+        optimizer=tf.keras.optimizers.Adam(config.lr),
+        loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+        metrics=["accuracy"],
+    )
+
+    mlflow.tensorflow.autolog()
+    model.fit(
+        x_train,
+        y_train,
+        epochs=config.epochs,
+    )
+
+    # write model
+    return model
+
+
+# Define the step and enable mlflow - order of decorators is important here
+@enable_mlflow
+@step
+def tf_evaluator(
+    x_test: np.ndarray,
+    y_test: np.ndarray,
+    model: tf.keras.Model,
+) -> float:
+    """Calculate the loss for the model for each epoch in a graph"""
+
+    _, test_acc = model.evaluate(x_test, y_test, verbose=2)
+    mlflow.log_metric("val_accuracy", test_acc)
+    return test_acc
+
+
+# Define the step and enable mlflow - order of decorators is important here
+@enable_mlflow
+@step(enable_cache=True)
+def predictor(
+    model: tf.keras.Model, context: StepContext
+) -> MLFlowDeploymentService:
+    """Start a prediction service
+
+    NOTE: the input argument is just a dummy that allows us to enforce that
+    the prediction step is executed after the training step, otherwise no
+    model is found in the current MLflow run.
+    """
+
+    def get_last_service() -> MLFlowDeploymentService:
+        """Get the service created during the last step run."""
+        # get the current pipeline name and step name
+        step_env = Environment()[STEP_ENVIRONMENT_NAME]
+        pipeline_name = step_env.pipeline_name
+        step_name = step_env.step_name
+
+        pipeline = context.metadata_store.get_pipeline(
+            pipeline_name=pipeline_name
+        )
+
+        if len(pipeline.runs) < 2:
+            return None
+        step = pipeline.runs[-2].get_step(name=step_name)
+        for artifact_view in step.outputs.values():
+            # filter out anything but service artifacts
+            if artifact_view.type == "ServiceArtifact":
+                return artifact_view.read()
+
+    last_service = get_last_service()
+
+    model_uri = mlflow.get_artifact_uri("model")
+    if not model_uri:
+        # an MLflow model was not found in the current run, so we simply reuse
```

Comment:
It shouldn't happen with this particular example, because an MLflow model is always trained during the first pipeline run and then there will always be a last_service in all subsequent pipeline runs. However, raising an exception can't hurt.